File: src/main/java/net/golem/raknet/protocol/unconnected/UnconnectedPongPacket.java
package net.golem.raknet.protocol.unconnected;

import net.golem.raknet.codec.PacketDecoder;
import net.golem.raknet.codec.PacketEncoder;
import net.golem.raknet.protocol.DataPacket;
import net.golem.raknet.protocol.RakNetPacketIds;

public class UnconnectedPongPacket extends DataPacket {

	public long pingId;

	public long guid;

	public String serverName;

	public UnconnectedPongPacket() {
		super(RakNetPacketIds.UNCONNECTED_PONG);
	}

	@Override
	public void encode(PacketEncoder encoder) {
		encoder.writeLong(pingId);
		encoder.writeLong(guid);
		encoder.writeMagic();
		encoder.writeString(serverName, PacketEncoder.SHORT_STRING);
	}

	@Override
	public void decode(PacketDecoder decoder) {
		pingId = decoder.readLong();
		guid = decoder.readLong();
		decoder.skipMagic();
		serverName = decoder.readString();
	}
}

File: src/main/kotlin/raknet/packet/protocol/UnconnectedPongPacket.kt
package raknet.packet.protocol

import io.netty.buffer.ByteBuf
import raknet.Magic
import raknet.packet.DataPacket
import raknet.packet.PacketType
import raknet.readMagic

class UnconnectedPongPacket constructor(
    var pingId: Long,
    var guid: Long,
    var magic: Magic = Magic,
    var serverName: String
    ): DataPacket(PacketType.UNCONNECTED_PONG.id()) {

    override fun decode(buffer: ByteBuf) {

    }

    companion object {
        fun from(buffer: ByteBuf): UnconnectedPongPacket {
            return UnconnectedPongPacket(
                buffer.readLong(),
                buffer.readLong(),
                buffer.readMagic(),
                buffer.readCharSequence(buffer.readShort().toInt(), Charsets.UTF_8).toString()
            )
        }
    }

    override fun toString(): String {
        return "UnconnectedPongPacket(pingId=$pingId, guid=$guid, serverName='$serverName')"
    }

}